agent/lib/kontena/workers/weave_worker.rb

```diff
@@ -16,17 +16,43 @@ def initialize
 
       @migrate_containers = nil # initialized by #start
 
+      # In case weave gets re-created (via version change, trusted-subnets, ...)
+      # it will lose all local DNS names thus they need to be re-added
+      subscribe('network_adapter:restart', :ensure_dns)
       if network_adapter.running?
         self.start
       else
         subscribe('network_adapter:start', :on_weave_start)
       end
+
     end
 
     def on_weave_start(topic, data)
       self.start
     end
 
+    def ensure_dns(topic, data)
+      info 'ensuring existing containers have proper DNS names'
```

Comment:
These `ensuring foobar` log lines are a bit uncomfortable because you see these in the logs but you never know if it did complete or not. So maybe add same kind of log line to the end that reports how many dns entries has been "ensured".